packages/text-input/src/TextInput.tsx

```diff
@@ -0,0 +1,337 @@
+      if (onChange) {
+        onChange(e);
+      }
+
+      if (!isControlled) {
+        setValue(e.target.value);
+      }
+    }
+
+    const interactionRingFocusStyle = css`
+      ${inputSelectorProp.selector}:focus ~ & {
+        background-color: #9dd0e7;
+        transform: scale(1);
+        z-index: 1;
+      }
+    `;
+
+    const interactionRingHoverStyle = css`
+      ${inputSelectorProp.selector}:hover ~ & {
+        transform: scale(1);
+      }
+    `;
+
+    const descriptionStyle = css`
+      color: ${uiColors.gray.dark1};
+      font-size: 14px;
+      height: ${description === '' ? '0' : '20'}px;
+      font-weight: normal;
+      padding-bottom: 8px;
+    `;
+
+    const inputStyle = css`
+      width: 400px;
+      height: 36px;
+      border-radius: 4px;
+      padding-left: 12px;
+      font-size: 14px;
+      font-weight: normal;
+      z-index: ${hasFocus ? 2 : 1};
+      background-color: ${disabled ? uiColors.gray.light2 : '#FFFFFF'};
+      padding-right: ${getInputPaddingFromState()};
+      border: 1px solid ${getInputColorFromState()};
+      &:placeholder {
+        color: ${uiColors.gray.base};
+      }
+      &:focus {
+        outline: none;
+      }
+    `;
+
+    return (
+      <label className={cx(textInputStyle, labelStyle, className)}>
+        {label}
+        <label className={descriptionStyle}>{description}</label>
+        <div className={inputContainerStyle}>
+          <input
+            {...inputSelectorProp.prop}
+            className={inputStyle}
+            type="text"
+            value={value}
+            required={!optional}
+            disabled={disabled}
+            placeholder={placeholder}
+            onChange={onValueChange}
+            ref={forwardRef}
+            onFocus={() => setHasFocus(true)}
+            onBlur={() => setHasFocus(false)}
+          />
+
+          {state === State.Valid && (
+            <Icon
+              glyph="Checkmark"
+              className={cx(
+                validIconStyle,
+                css`
+                  z-index: ${hasFocus ? 2 : 1};
+                `,
+              )}
+            />
+          )}
+
+          {state === State.Error && (
+            <Icon
+              glyph="Warning"
+              className={cx(
+                errorIconStyle,
+                css`
+                  z-index: ${hasFocus ? 2 : 1};
+                `,
+              )}
+            />
+          )}
+
+          {state === State.None && optional && (
+            <div
+              className={cx(
+                optionalStyle,
+                css`
+                  z-index: ${hasFocus ? 2 : 1};
+                `,
+              )}
+            >
+              <p>Optional</p>
+            </div>
+          )}
+
+          <div
+            className={cx(interactionRing, interactionRingHoverStyle, {
+              [interactionRingFocusStyle]: showFocus,
+            })}
+          />
+        </div>
+        {state === State.Error && (
```

Comment:
Should also check if `errorMessage` exists before we render this block?

Reply:
yes, I think so- unless there is a case where you would want the error state/icon to show but not an error message? but that sounds weird so I'm going to guess it wouldn't happen.